Add comprehensive tests for custom Array reduce methodUpdate reduce.t…

JoAnna578 opened the pull request:
…est.js

Dodano zestaw testów dla metody reduce2, obejmujący: sumowanie, mnożenie, brak wartości początkowej, pustą tablicę, oraz sprawdzenie poprawności argumentów przekazywanych do callbacka. Testy zapewniają pełną zgodność z zachowaniem natywnej metody Array.prototype.reduce.

JoAnna578 committed:
Add comprehensive tests for custom Array reduce methodUpdate reduce.test.js

Dodano zestaw testów dla metody reduce2, obejmujący: sumowanie, mnożenie, brak wartości początkowej, pustą tablicę, oraz sprawdzenie poprawności argumentów przekazywanych do callbacka. Testy zapewniają pełną zgodność z zachowaniem natywnej metody Array.prototype.reduce.

JoAnna578 committed:
Poprawa implementacji reduce – obsługa pustej tablicy bez wartości początkowej

Doprecyzowano działanie funkcji reduce, tak aby rzucała wyjątek, gdy tablica jest pusta i nie podano wartości początkowej. Dzięki temu wszystkie testy, w tym sprawdzające błędne przypadki, przechodzą poprawnie. Pozostałe przypadki redukcji pozostają bez zmian.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
+    const arr = [1, 2, 3];
+    const result = arr.reduce2((acc, val) => acc + val);
+    expect(result).toBe(6);
+  });
 
+  it('should handle empty array with initial value', () => {
+    const arr = [];
+    const result = arr.reduce2((acc, val) => acc + val, 5);
+    expect(result).toBe(5);
```
This test verifies the return value for an empty array with an initial value, which is required. However, it should also assert that the reducer callback is not called. Use a spy (e.g., `const cb = jest.fn(); arr.reduce2(cb, 5); expect(cb).toHaveBeenCalledTimes(0);`) to ensure the callback is not invoked, matching native `reduce` behavior.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
-  // Add tests here
+  it('should throw error when empty array and no initial value', () => {
+    const arr = [];
+    expect(() => arr.reduce2((acc, val) => acc + val)).toThrow();
```
The empty-array-without-initial-value test uses `toThrow()` generically. The native `Array.prototype.reduce` throws a `TypeError` in this case. Make this assertion specific by using `toThrow(TypeError)` so the custom implementation must match the native error type.

mateacademy-ai-mentor replied:
Where you currently assert thrown errors (line 40), prefer asserting the exact error type and message if applicable. This ensures parity with native behavior and prevents implementations that throw generic errors from passing.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
+    expect(result).toBe(24);
+  });
+
+  it('should work without initial value', () => {
+    const arr = [1, 2, 3];
+    const result = arr.reduce2((acc, val) => acc + val);
+    expect(result).toBe(6);
```
Add tests for single-element arrays with and without an initial value. For example: (1) `[5].reduce2((acc, v) => acc + v)` should return `5`; (2) `[5].reduce2((acc, v) => acc + v, 10)` should return `15`. These confirm correct behavior when using the first element as the initial accumulator and when an explicit initial is provided.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
   });
 
-  it('should ', () => {
+  it('should sum numbers in an array', () => {
+    const arr = [1, 2, 3, 4];
+    const result = arr.reduce2((acc, val) => acc + val, 0);
+    expect(result).toBe(10);
```
Include tests that exercise non-number types to ensure accumulator propagation works across types. For example, a string concatenation test (`['a','b','c'].reduce2((a,b)=>a+b, '')`) and an object-merge test (accumulating properties into an object) are required by the description.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
+    expect(result).toBe(10);
+  });
+
+  it('should multiply numbers in an array', () => {
+    const arr = [1, 2, 3, 4];
+    const result = arr.reduce2((acc, val) => acc * val, 1);
+    expect(result).toBe(24);
```
Add tests that verify mutation-during-reduction semantics. Specifically, check that elements appended after reduction starts are not visited (length captured at start) and elements removed before being visited are skipped. Create a test where the reducer mutates the array (push/splice/pop) and then assert which indices were visited and how many times the callback was called.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
+  });
+
+  it('should pass correct arguments to callback', () => {
+    const arr = [10, 20];
```
Add a test for sparse arrays (holes). For example: `const arr = [1, , 3]; const cb = jest.fn((acc, v, idx) => acc + v); arr.reduce2(cb, 0);` then assert `cb` was called for indices `0` and `2` only and that the hole at index `1` was skipped. This matches native `reduce` semantics and is required by the description.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
+
+    expect(mockCb).toHaveBeenCalledTimes(2);
+    expect(mockCb).toHaveBeenNthCalledWith(1, 0, 10, 0, arr);
+    expect(mockCb).toHaveBeenNthCalledWith(2, 10, 20, 1, arr);
```
Add tests for callback `this` behavior. The description requires verifying that `reduce`'s callback doesn't implicitly change `this`. For example, use an unbound function and a bound function to assert `this` inside the callback is as expected (or `undefined`/global depending on strict mode) and mirrors native behavior.

mateacademy-ai-mentor commented on src/reduce.test.js:
```diff
@@ -11,9 +11,42 @@ describe('reduce', () => {
-  it('should ', () => {
+  it('should sum numbers in an array', () => {
+    const arr = [1, 2, 3, 4];
+    const result = arr.reduce2((acc, val) => acc + val, 0);
```
Add an explicit test ensuring `reduce` itself does not modify the original array (unless the reducer deliberately mutates it). For example, copy the array before/after calling `reduce2` and assert the array content and length are unchanged when the reducer doesn't mutate the array.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I am requesting changes because several required edge-case tests are missing, which blocks meeting the task’s completeness criteria. You did a solid job covering core scenarios (sum/product, with/without initial, empty array behaviors, callback arguments), but the suite must also match native `reduce` semantics across edge cases.

Critical issues to fix:
- Add a spy-based assertion that the reducer is not called for an empty array when an initial value is provided (e.g., around current test near line 30). 
- For empty array without an initial value (around line 40), assert `toThrow(TypeError)` and, if applicable, the native message.
- Add tests for single-element arrays: without initial returns the element and does not call the callback; with initial processes once and returns the correct sum.
- Add a test that non-function callbacks throw `TypeError`.
- Add tests for sparse arrays (holes) to ensure skipped indices and correct call indices.
- Add tests for mutation-during-reduction (push/splice/pop) to confirm length captured at start and correct visitation.
- Add tests for non-number types (string concatenation, object accumulation) to verify accumulator propagation.
- Add tests for callback `this` behavior (unbound vs bound) to mirror native behavior.
- Add a test ensuring `reduce2` does not modify the original array when the reducer doesn’t mutate it.

Minor improvements: also consider a test where the initial value is explicitly `undefined` to ensure it’s treated as provided and invokes the callback. Once these are added, your suite will comprehensively validate parity with native `reduce`. Keep up the good work—your current structure makes adding these cases straightforward.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **No changes were made in the pull request since the last review.** Please make sure that you have added files to the pull request and committed the changes.
> <sub>*[CODE: NCM]*</sub>


If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in about 1 hour.
If you believe this is an error or need assistance, please contact the support team.

JoAnna578 committed:
Finalna wersja qa array

EwaKobiela reviewed (changes requested):
Please introduce changes suggested by AI mentor

JoAnna578 committed:
Add comprehensive tests for custom reduce2 implementation

Rozszerzono zestaw testów reduce2 o przypadki krawędziowe: pusty array, single-element array, sparse arrays, mutacje w trakcie redukcji, różne typy danych, poprawne wywołanie callback oraz obsługę błędów. Teraz reduce2 zachowuje się zgodnie z natywnym Array.prototype.reduce.

JoAnna578 committed:
Add comprehensive tests for custom reduce2 implementation 2

JoAnna578 committed:
Poprawki testów reduce2 – linie i tablice nieciągłe

Poprawiłam długie linie w testach toHaveBeenNthCalledWith, zastąpiłam dziury w tablicach jawnym undefined, dodałam sprawdzenie, że callback nie jest wywoływany dla pustej tablicy z wartością początkową.

JoAnna578 committed:
Poprawki testów reduce2 – linie i tablice nieciągłe 2
